Sources/PipeableSDK/PipeablePage.swift

```diff
@@ -496,6 +524,113 @@ public struct XHRResult: Decodable {
     public var responseType: ResponseType
 }
 
+enum LoadState: Int {
+    case notloaded = 0
+    case domcontentloaded = 1
+    case load = 2
+    case networkidle = 3
+
+    static func fromWaitUntil(_ waitUntilOption: PipeablePage.WaitUntilOption) -> LoadState {
+        switch waitUntilOption {
+        case .load:
+            return .load
+        case .domcontentloaded:
+            return .domcontentloaded
+        case .networkidle:
+            return .networkidle
+        }
+    }
+
+    static func fromString(_ stringvalue: String) -> LoadState? {
+        switch stringvalue {
+        case "domcontentloaded":
+            return .domcontentloaded
+        case "load":
+            return .load
+        case "networkidle":
+            return .networkidle
+        default:
+            return nil
+        }
+    }
+}
+
+class PageLoadState {
+    // We start in a not loaded state and then go through the different stages.
+    // They are determined through the delegate for domcontentloaded and then
+    // from javascript on the page itself for the other types.
+    private(set) var state: LoadState = .notloaded
+    private var loadError: Error?
+    private var currentURL: String?
+    private var callbacks: [CallbackWrapper] = []
+
+    private let synchronizationQueue = DispatchQueue(label: "waitForURL")
+
+    // Define a struct to wrap the callback and its unique identifier
+    private class CallbackWrapper: Equatable {
+        let id: UUID = .init()
+        let callback: LoadStateChangeCallback
+
+        init(callback: @escaping LoadStateChangeCallback) {
+            self.callback = callback
+        }
+
+        static func == (lhs: PageLoadState.CallbackWrapper, rhs: PageLoadState.CallbackWrapper) -> Bool {
```

Comment:
Why do you need the UUID and the equal override? Doesn't `===` (comparison by ref) work just fine?